src/main/java/com/linecorp/armeria/server/thrift/xml/ThriftXmlDocString.java

```diff
@@ -0,0 +1,195 @@
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.server.thrift.xml;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Deque;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import javax.xml.stream.XMLInputFactory;
+import javax.xml.stream.XMLStreamException;
+import javax.xml.stream.XMLStreamReader;
+import javax.xml.stream.events.XMLEvent;
+
+import com.google.common.collect.Sets;
+import com.google.common.reflect.ClassPath;
+import com.google.common.reflect.ClassPath.ClassInfo;
+
+import io.netty.util.internal.ObjectUtil;
+
+/**
+ * {@link ThriftXmlDocString} is an DocString extractor for Thrift IDL XML.
+ */
+public final class ThriftXmlDocString {
+
+    private static final ClassLoader CLASS_LOADER = ThriftXmlDocString.class.getClassLoader();
+
+    private static final Set<String> DOC_CONTAINABLE_XML_ELEMENTS =
+            Sets.newHashSet("enum", "member", "struct", "method", "field", "arg", "exception", "service",
+                            "throws");
+
+    private static final XMLInputFactory XML_INPUT_FACTORY = XMLInputFactory.newInstance();
+
+    private static final String CONSTANTS_SUFFIX = "Constants";
+
+    private ThriftXmlDocString() {
+    }
+
+    /**
+     * Extracts all DocStrings from all Thrift IDL XMl Resources.
+     * @return a map with key is FQCN and value is document string.
+     */
+    public static Map<String, String> getAllDocStrings() {
+        return parseDocStrings(getAllThriftXmls());
+    }
+
+    /**
+     * Parses DocStrings from input Thrift IDL XMl Resources.
+     * @return a map with key is FQCN and value is document string.
+     */
+    public static Map<String, String> parseDocStrings(List<String> xmlPaths) {
+        Map<String, String> docStrings = new HashMap<>();
+        for (String xmlPath : xmlPaths) {
+            Map<String, String> values = getDocStringsFromXmlResource(xmlPath);
+            if (values != null) {
+                docStrings.putAll(values);
+            }
+        }
+        return docStrings;
+    }
+
+    /**
+     * Gets all thrift XMLs from current class loader.
+     * @return a list with xml paths.
+     */
+    public static List<String> getAllThriftXmls() {
+        List<String> xmlPaths = new ArrayList<>();
+        try {
+            Set<ClassInfo> topLevelClasses = ClassPath.from(CLASS_LOADER).getTopLevelClasses();
```

Comment:
Yeah, but otherwise this will fail to find the files in a multi-classloader environment. We could avoid scanning for the same class loader multiple times though when more than 1 service were loaded by the same class loader.